pkg/kubevirt/utils.go

```diff
@@ -138,36 +152,70 @@ func buildVirtualMachineInstanceTemplate(ctx *context.MachineContext) *kubevirtv
 
 	template.Spec = *ctx.KubevirtMachine.Spec.VirtualMachineTemplate.Spec.Template.Spec.DeepCopy()
 
-	cloudInitVolumeName := "cloudinitvolume"
-	cloudInitVolume := kubevirtv1.Volume{
-		Name: cloudInitVolumeName,
-		VolumeSource: kubevirtv1.VolumeSource{
-			CloudInitConfigDrive: &kubevirtv1.CloudInitConfigDriveSource{
-				UserDataSecretRef: &corev1.LocalObjectReference{
-					Name: *ctx.Machine.Spec.Bootstrap.DataSecretName + "-userdata",
+	cloudInitType, index := detectCloudInitVolumeType(template)
+	cloudInitVolume := cloudinitVolume(template, ctx, cloudInitType)
+	cloudInitDisk := cloudinitDisk(template)
+
+	if index >= 0 {
+		template.Spec.Volumes[index] = cloudInitVolume
+	} else {
+		template.Spec.Domain.Devices.Disks = append(template.Spec.Domain.Devices.Disks, cloudInitDisk)
+	}
```

Comment:
Is this what you are looking for @davidvossel ?

Reply:
not exactly.

The disk and volume logic needs to be independent. Here's what we need.

If the volume already exists, override it `template.Spec.Volumes[index] = cloudInitVolume` else append the volume to the template.Spec.Volumes slice.

If the disk matching the volume name already exists, then leave it alone. If the disk doesn't exist, append it.

The idea here is that someone can create their own disk mapping to the cloud init volume, using whatever parameters they want and we'll use that... but if no disk is provided that matches the cloud-init volume name, we'll automatically create one using the virtio bus.

And for the volumes, some one can create a volume with a specific cloud init type, and we'll inject the user data into that, or if no volume exists already for cloud init, we'll create one for the user defaulting to cloud init config drive.